x/axelarnet/message_handler.go

```diff
@@ -12,22 +12,84 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 
 	"github.com/axelarnetwork/axelar-core/utils/events"
+	axelarnet "github.com/axelarnetwork/axelar-core/x/axelarnet/exported"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/keeper"
 	"github.com/axelarnetwork/axelar-core/x/axelarnet/types"
 	evmtypes "github.com/axelarnetwork/axelar-core/x/evm/types"
 	nexus "github.com/axelarnetwork/axelar-core/x/nexus/exported"
 	"github.com/axelarnetwork/utils/funcs"
 )
 
+// Fee is used to pay relayer for executing cross chain message
+type Fee struct {
+	Amount    string `json:"amount"`
+	Recipient string `json:"recipient"`
+}
+
 // Message is attached in ICS20 packet memo field
 type Message struct {
 	DestinationChain   string `json:"destination_chain"`
 	DestinationAddress string `json:"destination_address"`
 	Payload            []byte `json:"payload"`
 	Type               int64  `json:"type"`
+	Fee                *Fee   `json:"fee"` // Optional
+}
+
+func (f Fee) validate(ctx sdk.Context, n types.Nexus, b types.BankKeeper, token sdk.Coin, msgType nexus.MessageType) error {
+	amt, ok := sdk.NewIntFromString(f.Amount)
+	if !ok || amt.LTE(sdk.ZeroInt()) {
+		return fmt.Errorf("invalid fee amount")
+	}
+
+	afterFee := token.Amount.Sub(amt)
+	switch msgType {
+	case nexus.TypeGeneralMessage:
+		if afterFee.LT(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater than transfer value")
+		}
+	case nexus.TypeGeneralMessageWithToken:
+		if afterFee.LTE(sdk.ZeroInt()) {
+			return fmt.Errorf("fee amount is greater or equal to transfer value")
+		}
```

Comment:
why the different checks?

Reply:
contract call with token require non zero amount after fee